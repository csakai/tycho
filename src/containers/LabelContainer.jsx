import React from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import * as Actions from '../actions/LabelActions';
import ReduxService from '../services/ReduxService';
import Label from '../components/Label';
import Constants from '../constants';

export class LabelContainer extends React.Component {

  static propTypes = {
    position: PropTypes.object,
    text: PropTypes.string.isRequired,
    id: PropTypes.string.isRequired
  }

  /**
   * Sets the targetName global state.
   */
  setActiveOrbital = () => {
    this.props.action.setActiveOrbital(this.props.id);
    this.props.action.setLabelText(this.props.text);
  }

  /**
   * Sets the currently-highlighted orbital path.
   */
  setHighlightedOrbital = () => {
    this.props.action.setHighlightedOrbital(this.props.id);
  }

  /**
   * Removes the currently-highlighted orbital path.
   */
  removeHighlightedOrbital = () => {
    this.props.action.setHighlightedOrbital(null);
  }

  /**
   * Determine if the current label is active or not.
   *
   * @returns {Boolean} state of label
   */
  isActive = () => {
    const {targetName, id, zoom} = this.props;
    return targetName === id && zoom <= Constants.UI.ZOOM_LABEL_TRIGGER;
  }

  /**
   * Returns the position of the id, if any defined.
   *
   * @returns {Object} 2d position of object
   */
  getPosition = () => {
    const {positions, id} = this.props;

    if (positions && positions[id]) {
      return positions[id].position2d;
    }
  }
  
  render() {
    return (
      <Label
        position={this.getPosition()}
        onClick={this.setActiveOrbital}
        onMouseOver={this.setHighlightedOrbital}
        onMouseOut={this.removeHighlightedOrbital}
        active={this.isActive()}
        enabled={this.props.controlsEnabled}
        text={this.props.text}>
        {this.props.children}
      </Label>
    );
  }
}

export default connect(
  ReduxService.mapStateToProps(
    'label.targetName',
    'uiControls.zoom',
    'uiControls.controlsEnabled',
    'animation.positions'
  ),
  ReduxService.mapDispatchToProps(Actions)
)(LabelContainer);
