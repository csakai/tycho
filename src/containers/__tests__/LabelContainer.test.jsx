import React from 'react';
import toJson from 'enzyme-to-json';
import {shallow} from 'enzyme';
import {LabelContainer} from '../LabelContainer';

describe('Label Container', () => {
  let component, labelContainer;

  beforeEach(() => {
    component = shallow(<LabelContainer
      text="foo"
      positions={{}}
      id="testPlanet"
    />);

    labelContainer = component.instance();
  });

  describe('getPosition()', () => {
    it('should return a style object with top, left, and position definitions', () => {
      const pos = {top: 4, left: 10};
      const style = labelContainer.getPosition(pos);

      expect(typeof style).toBe('object');
      expect(style).toHaveProperty('position');
      expect(style).toHaveProperty('top');
      expect(style).toHaveProperty('left');
      expect(style.position).toEqual('absolute');
      expect(style.top).toEqual(`${pos.top}px`);
      expect(style.left).toEqual(`${pos.left}px`);
    });

    it('should return null if the position parameter is not defined', () => {
      const style = labelContainer.getPosition();

      expect(style).toBeDefined();
      expect(style).toBeNull();
    });
  });

  describe('getChildren()', () => {
    const children = [];
    
    it('should return `children` when active', () => {
      const result = labelContainer.getChildren(children, true);
      expect(result).toEqual(children);
    });

    it('should return null when not active', () => {
      const result = labelContainer.getChildren(children);
      expect(result).toEqual(null);
    });
  });

  describe('setActiveOrbital()', () => {
    it('should call the setActiveOrbital action with the id', () => {
      const setActiveOrbital = jest.fn();
      const id = 'testPlanet';

      labelContainer.props = {
        action: {setActiveOrbital},
        id
      };
      const spy = jest.spyOn(labelContainer.props.action, 'setActiveOrbital');

      labelContainer.setActiveOrbital();

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledWith(id);
    });
  });

  describe('setHighlightedOrbital()', () => {
    it('should call the setActiveOrbital action with the id', () => {
      const setHighlightedOrbital = jest.fn();
      const id = 'testPlanet';

      labelContainer.props = {
        action: {setHighlightedOrbital},
        id
      };
      const spy = jest.spyOn(labelContainer.props.action, 'setHighlightedOrbital');

      labelContainer.setHighlightedOrbital();

      expect(spy).toHaveBeenCalled();
      expect(spy).toHaveBeenCalledWith(id);
    });
  });

  describe('render()', () => {
    it('should render the Label successfully', () => {
      expect(toJson(component)).toMatchSnapshot();
    });
  });
});
