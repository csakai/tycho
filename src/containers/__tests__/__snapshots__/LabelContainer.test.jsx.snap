// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label Container render() should render the Label successfully 1`] = `
<Label
  onClick={[Function]}
  onHover={[Function]}
  text="foo"
/>
`;
